python/ray/data/datasource/tfrecords_datasource.py

```diff
@@ -64,26 +64,19 @@ def _write_block(
 
 def _convert_example_to_dict(
     example: "tf.train.Example",
-) -> Dict[
-    str,
-    Union[
-        List[bytes],
-        List[List[bytes]],
-        List[float],
-        List[List[float]],
-        List[int],
-        List[List[int]],
-    ],
-]:
+) -> Dict[str, "pyarrow.Array"]:
+    import pyarrow as pa
+
     record = {}
     for feature_name, feature in example.features.feature.items():
         value = _get_feature_value(feature)
+        value = value.to_pylist()
         # Return value itself if the list has single value.
         # This is to give better user experience when writing preprocessing UDF on
         # these single-value lists.
         if len(value) == 1:
             value = value[0]
-        record[feature_name] = [value]
+        record[feature_name] = pa.array([value])
```

Comment:
Could we push this logic into `_get_feature_value()`, now that returning an Arrow `Array`? ideally this could just be `record[feature_name] = _get_feature_value(feature)`